cache/memoize.ts

```diff
@@ -0,0 +1,203 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+import { LruCache } from "./lru_cache.ts";
+
+export type MemoizationCache<K, V> = {
+  has: (key: K) => boolean;
+  get: (key: K) => V | undefined;
+  set: (key: K, val: V) => unknown;
+  delete: (key: K) => unknown;
+};
+
+export type MemoizationOptions<Fn extends (...args: never[]) => unknown> = {
+  /**
+   * Provide a custom cache for getting previous results. By default, a new
+   * `Map` object is instantiated upon memoization and used as a cache.
+   */
+  cache?: MemoizationCache<unknown, ReturnType<Fn>>;
+  /**
+   * Function to get a unique cache key from the function's arguments. By
+   * default, a composite key is created from all the arguments plus the `this`
+   * value, using reference equality to check for equivalence.
+   *
+   * @example
+   * ```ts
+   * import { memoize } from "@std/cache";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const fn = memoize(({ value }: { cacheKey: number; value: number }) => {
+   *   return value;
+   * }, { getKey: ({ cacheKey }) => cacheKey });
+   *
+   * assertEquals(fn({ cacheKey: 1, value: 2 }), 2);
+   * assertEquals(fn({ cacheKey: 1, value: 99 }), 2);
+   * assertEquals(fn({ cacheKey: 2, value: 99 }), 99);
+   * ```
+   */
+  getKey?: (this: ThisParameterType<Fn>, ...args: Parameters<Fn>) => unknown;
+  /**
+   * Only use args as cache keys up to the `length` property of the function.
+   * Useful for passing unary functions as array callbacks, but should be
+   * avoided for functions with variable argument length (`...rest` or default
+   * params)
+   *
+   * @default false
+   */
+  truncateArgs?: boolean;
+  /**
+   * By default, promises are automatically removed from the cache upon
+   * rejection. If `cacheRejectedPromises` is set to `true`, promises will be
+   * retained in the cache even if rejected.
+   *
+   * @default false
+   */
+  cacheRejectedPromises?: boolean;
+};
+
+// used for memoizing the `memoize` function itself to enable on-the-fly usage
```

Comment:
What is the `on-the-fly` usage?

memoize function being memoized feels surprising to me.

Reply:
Ah, thanks (I didn't read the PR description 😓 

I personally don't expect `memoize(fn)(42)` works in a memoized way. Do you know any prior example which supports that pattern, or is this just invented?

Reply:
Not aware of a prior example, it just seemed like a nifty feature to have. But I think on balance the small bump in complexity coupled with the gotcha are reason enough to remove it.